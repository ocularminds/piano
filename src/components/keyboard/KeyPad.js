import React from 'react';
import './Keyboard.css';

class KeyPad extends React.Component {
  constructor(props) {
    super(props);
    this.createPad = this.createPad.bind(this);
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick(e) {
    e.preventDefault();
    if (e.target.getAttribute('label') && this.props.clazz === 'WHITE') {
      e.target.style.setProperty('background-color', '#737373');
      this.props.onSave(e.target.getAttribute('label'));
    }
  }

  createPad() {
    const pad = this.props.pad;
    let zIndex = parseInt(this.props.id) + 1;
    let top = parseInt(this.props.id) === 0 ? 0 : -200;
    let left = Math.ceil(parseInt(this.props.id) * 52.5);
    if (this.props.clazz === 'BLACK') {
      zIndex = parseInt(this.props.id) + 1 + parseInt(this.props.size);
      top = parseInt(this.props.id) === 0 ? -200 : -150;
      left = Math.ceil(parseInt(this.props.id) * 52.5) + 26;
    }
    const semitone = pad === 'E' || pad === 'B';
    const label = this.props.clazz === 'WHITE' ? pad : pad + '#';
    const stylePads = {
      marginLeft: `${left}px`,
      zIndex: `${zIndex}`,
      marginTop: `${top}px`,
    };
    return {stylePads, semitone, label};
  }
  render() {
    const key = this.createPad();
    return (
      <div
        className={this.props.clazz.toLowerCase()}
        style={key.stylePads}
        label={key.label}
        onClick={(e) => this.handleClick(e)}
      >
        <div className="key">{key.label}</div>
      </div>
    );
  }
}

export default KeyPad;
