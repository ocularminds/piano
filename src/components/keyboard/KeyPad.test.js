import React from 'react';
import {render} from '@testing-library/react';
import KeyPad from './KeyPad';
import {shallow, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({adapter: new Adapter()});

describe('Test KeyPad components when rendering', () => {
  it('should match the snapshot', () => {
    const save = (e) => {};
    const wrapper = shallow(
      <KeyPad pad={'F'} clazz={'WHITE'} keyid={3} onSave={save} id={1} />
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('renders keypad with label', () => {
    const save = (e) => {};
    const {getByText} = render(
      <KeyPad pad={'F'} clazz={'WHITE'} keyid={3} onSave={save} id={1} />
    );
    const linkElement = getByText(/F/i);
    expect(linkElement).toBeInTheDocument();
  });

  it('Should contain piano name in the properties', () => {
    const save = (e) => {};
    const wrapper = shallow(
      <KeyPad pad={'F'} clazz={'WHITE'} keyid={3} onSave={save} id={1} />
    );
    const result = wrapper.instance().createPad();
    expect(result['semitone']).toEqual(false);
  });

  it('Should contain black key with # surfix', () => {
    const save = (e) => {};
    const wrapper = shallow(
      <KeyPad pad={'A'} clazz={'BLACK'} keyid={3} onSave={save} id={1} />
    );
    const result = wrapper.instance().createPad();
    expect(result['semitone']).toEqual(false);
  });

  it('Should not contain black key for semitones', () => {
    const save = (e) => {};
    const wrapper = shallow(
      <KeyPad pad={'B'} clazz={'BLACK'} keyid={3} onSave={save} id={1} />
    );
    const result = wrapper.instance().createPad();
    expect(result['semitone']).toEqual(true);
  });
});
