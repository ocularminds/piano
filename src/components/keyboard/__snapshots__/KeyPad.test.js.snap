// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test KeyPad components when rendering should match the snapshot 1`] = `ShallowWrapper {}`;
